### PR TITLE
CL-25818, Upgraded geocoder version and fixed calculate_route, lookup and autocomplete API

### PR DESCRIPTION
**Pull request by harshaljaitpal**
https://jira.dev.e2open.com/jira/browse/CL-25818

### CODE REVIEW

**Comment by priyank-gupta** [comment]
@arpit1094 Please review.

File: lib/geocoder/lookups/here_calculate_route.rb [hunk 1]
```diff
@@ -26,15 +26,12 @@ def results(query)
...
-      )
+      super.merge(query_url_here_options(query))
     end
 
-    def waypoints_hash(waypoints)
-      waypoints.each_with_object({}).with_index do |(point, hash), index|
-        hash["waypoint#{index}"] = point.join(',')
-      end
+    def hash_to_query(hash)
+      URI.encode_www_form(hash)
```
Comment (arpit1094) [inline lib/geocoder/lookups/here_calculate_route.rb:33]: How is this ensuring waypoint are converted to format specified here:
https://developer.here.com/documentation/routing-api/dev_guide/topics/use-cases/via.html


We use this in HereMaps::ApiHelper#get_transport_report

File: lib/geocoder/lookups/here_lookup_id.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+module Geocoder::Lookup
+  class HereLookupId < Base
+    include HereLookupDefaultMethods
+
+    def name
+      "HereLookupId"
```
Comment (arpit1094) [inline lib/geocoder/lookups/here_lookup_id.rb:6]: Where are we using this lookup ?

File: lib/geocoder/lookups/here_suggest.rb [hunk 3]
```diff
@@ -25,9 +25,18 @@ def results(query)
       end
     end
 
+    def query_url_here_options(query)
```
Comment (arpit1094) [inline lib/geocoder/lookups/here_suggest.rb]: `query_url_here_options` is to be used for default url options which are common across all lookups, for eg `apiKey` and `lang`. Can we remove this override and move logic to `query_url_params` ?

File: lib/geocoder_here_maps/here.rb [hunk 4]
```diff
@@ -2,12 +2,12 @@
...
   Here.class_eval do
-    include HereLookupDefaultMethods
+    # include HereLookupDefaultMethods
 
     private
 
-    def base_query_url(query)
-      "#{protocol}://#{if query.reverse_geocode? then 'reverse.' end}geocoder.#{domain}/6.2/#{if query.reverse_geocode? then 'reverse' end}geocode.json?"
-    end
+    # def base_query_url(query)
```
Comment (arpit1094) [inline lib/geocoder_here_maps/here.rb]: Please remove the code if no longer required.

File: lib/geocoder/lookups/here_suggest.rb (hunk 3 above)
Comment (arpit1094) [reply lib/geocoder/lookups/here_suggest.rb]: Based on geocoder gem's latest code https://github.com/alexreisner/geocoder/blob/master/lib/geocoder/lookups/here.rb#L46 they have it for both `geocode` and `revgeocode`, if this option is supported for all apis let's move it to `HereLookupDefaultMethods#query_url_here_options` else to  `HereSuggest#query_url_params`

File: lib/geocoder/lookups/here_calculate_route.rb (hunk 1 above)
Comment (harshaljaitpal) [reply lib/geocoder/lookups/here_calculate_route.rb:33]: `waypoints_hash` method is no longer required. Waypoints can now be directly sent as params from request itself. No processing is required on waypoints params, that is why I removed it. `hash_to_query` method is not a replacement for `waypoints_hash` method. hash_to_query method ensures that hash is properly encoded. For eg: 
when we declare hash with key via which is an array,`a = {via: ['abc', 'ABC']}` and params should be encoded in the format as:`via=abc&via=ABC`, then we need to use the method encode_www_form.

File: lib/geocoder/lookups/here_lookup_id.rb (hunk 2 above)
Comment (harshaljaitpal) [reply lib/geocoder/lookups/here_lookup_id.rb:6]: Now, search by `location_id` is no longer supported by geocoder API. We need to use a new API request called as `/lookup` which is used to search by `reference_id/location_id`. Hence in order to make [this](https://github.com/cloudlogistics/cloudlogistics/blob/master/cl_system/app/lib/here_maps/api_helper.rb#L10) method working, I had to add new API support search by `lookup_id`. As this is not provided in geocoder gem as well. For more details, you can check [here](https://developer.here.com/documentation/geocoding-search-api/migration_guide/migration-geocoder/topics-api/geocoding.html#currently-not-supported) under the list of not supported things, look for locationid.

**Review by harshaljaitpal — COMMENTED** [review]
@arpit1094 I have resolved the comments. Can you please review again ?

**Comment by harshaljaitpal** [comment]
@arpit1094 Can you please review this ?

**Review by arpit1094 — COMMENTED** [review]
LGTM 👍 . Will approve finally once cloudlogistics repo PRs are finalized.

cc @harshaljaitpal  @priyank-gupta

**Review by arpit1094 — APPROVED** [review]
LGTM 👍

**Review by priyank-gupta — APPROVED** [review]
LGTM

**Comment by sergioloza** [comment]
ha!  I didnt even know about this repo

File: lib/geocoder/results/concerns/flexible_polyline_decoder.rb [hunk 5]
```diff
@@ -0,0 +1,97 @@
+module Geocoder::Result
```
Comment (sergioloza) [inline lib/geocoder/results/concerns/flexible_polyline_decoder.rb:1]: where did this logic come from? was it from this repo?https://github.com/heremaps/flexible-polyline/

Comment (harshaljaitpal) [reply lib/geocoder/results/concerns/flexible_polyline_decoder.rb:1]: Yes @sergioloza

**Pull request merged**